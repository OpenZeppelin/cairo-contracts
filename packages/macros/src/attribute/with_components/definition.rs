use cairo_lang_formatter::format_string;
use cairo_lang_macro::{attribute_macro, Diagnostic, ProcMacroResult, TokenStream};
use cairo_lang_parser::utils::SimpleParserDatabase;
use regex::Regex;

use super::{components::AllowedComponents, parser::WithComponentsParser};

/// Inserts multiple component dependencies into a modules codebase.
#[attribute_macro]
pub fn with_components(attribute_stream: TokenStream, item_stream: TokenStream) -> ProcMacroResult {
    let args = parse_args(&attribute_stream.to_string());

    // 1. Get the components info (if valid)
    let mut components_info = vec![];
    let empty_result = ProcMacroResult::new(TokenStream::empty());
    for arg in args {
        let maybe_component = AllowedComponents::from_str(&arg);
        match maybe_component {
            Ok(component) => {
                components_info.push(component.get_info());
            }
            Err(err) => {
                return empty_result.with_diagnostics(err.into());
            }
        }
    }

    // 2. Parse the item stream
    let db = SimpleParserDatabase::default();
    let (content, diagnostics) = match db.parse_virtual(item_stream.to_string()) {
        Ok(node) => WithComponentsParser::new(node, &components_info).parse(&db),
        Err(err) => {
            let error = Diagnostic::error(err.format(&db));
            return empty_result.with_diagnostics(error.into());
        }
    };

    let formatted_content = if !content.is_empty() {
        format_string(&db, content)
    } else {
        content
    };

    ProcMacroResult::new(TokenStream::new(formatted_content)).with_diagnostics(diagnostics)
}

/// Parses the arguments from the attribute stream.
fn parse_args(text: &str) -> Vec<String> {
    let re = Regex::new(r"(\w+)").unwrap();
    let matches = re.find_iter(text);
    matches.map(|m| m.as_str().to_string()).collect()
}

#[cfg(test)]
mod tests {
    use super::parse_args;

    #[test]
    fn test_parse_args() {
        let attribute = "(ERC20, Ownable)";
        let result = parse_args(attribute);
        assert_eq!(result, vec!["ERC20", "Ownable"]);

        let attribute = "ERC20";
        let result = parse_args(attribute);
        assert_eq!(result, vec!["ERC20"]);

        let attribute = "(Ownable, ERC20, Other, Another)";
        let result = parse_args(attribute);
        assert_eq!(result, vec!["Ownable", "ERC20", "Other", "Another"]);
    }
}
