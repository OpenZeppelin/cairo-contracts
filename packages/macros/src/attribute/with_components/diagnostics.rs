//! List of errors and warnings for the with_components attribute.

#[allow(non_snake_case)]
pub mod errors {
    /// Error when the module has no body.
    pub const NO_BODY: &str = "Contract module must have a body.\n";

    /// Error when the component is invalid.
    pub fn INVALID_COMPONENT(short_name: &str) -> String {
        format!("{} is not in the list of allowed components.\n", short_name)
    }
    /// Error when the module has no `#[starknet::contract]` attribute.
    pub fn NO_CONTRACT_ATTRIBUTE(contract_attribute: &str) -> String {
        format!(
            "Contract module must have the `#[{}]` attribute.\n",
            contract_attribute
        )
    }
}

#[allow(non_snake_case)]
pub mod warnings {
    use indoc::{formatdoc, indoc};

    /// Warning when the initializers for the components are missing.
    pub fn INITIALIZERS_MISSING(components: &str) -> String {
        formatdoc!(
            "It looks like the initializers for the following components are missing:

            {components}

            This may lead to unexpected behavior.
            We recommend adding the corresponding initializer calls to the constructor.
            "
        )
    }

    /// Warning when the component has no immutable config.
    pub fn IMMUTABLE_CONFIG_MISSING(short_name: &str, default_config_path: &str) -> String {
        formatdoc!(
            "The {} component requires an ImmutableConfig implementation in scope and
            it looks like it is missing.

            You can use the default implementation by importing it:

            `use {};`
            ",
            short_name,
            default_config_path
        )
    }

    /// Warning when the Vesting component is missing an implementation of the VestingScheduleTrait.
    pub const VESTING_SCHEDULE_IMPL_MISSING: &str = indoc! {
        "The Vesting component requires an implementation of the VestingScheduleTrait in scope and
        it looks like it is missing.

        You can use the LinearVestingSchedule implementation by importing it:

        `use openzeppelin_finance::vesting::LinearVestingSchedule;`
        "
    };

    /// Warning when the Initializable component is not used.
    pub const INITIALIZABLE_NOT_USED: &str = indoc! {
        "It looks like the `self.initializable.initialize()` function is not used in the contract. If
        this is intentional, you may consider removing the Initializable component.
        "
    };

    /// Warning when the Pausable component is not used.
    pub const PAUSABLE_NOT_USED: &str = indoc! {
        "It looks like the `self.pausable.pause()` and `self.pausable.unpause()` functions are not used in the contract. If
        this is intentional, you may consider removing the Pausable component.
        "
    };

    /// Warning when the ERC20 component is missing an implementation of the ERC20HooksTrait.
    pub const ERC20_HOOKS_IMPL_MISSING: &str = indoc! {
        "The ERC20 component requires an implementation of the ERC20HooksTrait in scope and
        it looks like it is missing.

        You can use the ERC20HooksEmptyImpl implementation by importing it:

        `use openzeppelin_token::erc20::ERC20HooksEmptyImpl;`
        "
    };

    /// Warning when the ERC721 component is missing an implementation of the ERC721HooksTrait.
    pub const ERC721_HOOKS_IMPL_MISSING: &str = indoc! {
        "The ERC721 component requires an implementation of the ERC721HooksTrait in scope and
        it looks like it is missing.

        You can use the ERC721HooksEmptyImpl implementation by importing it:

        `use openzeppelin_token::erc721::ERC721HooksEmptyImpl;`
        "
    };

    /// Warning when the ERC1155 component is missing an implementation of the ERC1155HooksTrait.
    pub const ERC1155_HOOKS_IMPL_MISSING: &str = indoc! {
        "The ERC1155 component requires an implementation of the ERC1155HooksTrait in scope and
        it looks like it is missing.

        You can use the ERC1155HooksEmptyImpl implementation by importing it:

        `use openzeppelin_token::erc1155::ERC1155HooksEmptyImpl;`
        "
    };

    /// Warning when the Upgradeable component is not used.
    pub const UPGRADEABLE_NOT_USED: &str = indoc! {
        "It looks like the `self.upgradeable.upgrade(new_class_hash)` function is not used in the contract. If
        this is intentional, you may consider removing the Upgradeable component.
        "
    };

    /// Warning when the Votes component is missing an implementation of the SNIP12Metadata trait.
    pub const SNIP12_METADATA_IMPL_MISSING: &str = indoc! {
        "The Votes component requires an implementation of the SNIP12Metadata trait in scope and
        it looks like it is missing.
        "
    };
}
