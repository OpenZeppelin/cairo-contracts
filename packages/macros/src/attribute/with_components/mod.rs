pub mod components;
pub mod definition;
pub mod diagnostics;
pub mod parser;
