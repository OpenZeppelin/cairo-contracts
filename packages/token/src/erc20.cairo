pub mod erc20;
pub mod interface;
pub mod snip12_utils;

pub use erc20::{ERC20Component, ERC20HooksEmptyImpl};
pub use interface::{ERC20ABIDispatcher, ERC20ABIDispatcherTrait};
