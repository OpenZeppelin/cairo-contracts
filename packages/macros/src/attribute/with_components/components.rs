use super::diagnostics::errors;
use cairo_lang_macro::Diagnostic;

/// The list of allowed components for the `with_components` attribute.
pub enum AllowedComponents {
    Account,
    EthAccount,
    SRC9,
    AccessControl,
    Ownable,
    Vesting,
    SRC5,
    Initializable,
    Pausable,
    ReentrancyGuard,
    ERC20,
    ERC721,
    ERC721Enumerable,
    ERC721Receiver,
    ERC1155,
    ERC1155Receiver,
    ERC2981,
    ERC4626,
    Upgradeable,
    Nonces,
    Multisig,
    TimelockController,
    Votes,
    Governor,
    GovernorCoreExecution,
    GovernorCountingSimple,
    GovernorSettings,
    GovernorTimelockExecution,
    GovernorVotesQuorumFraction,
    GovernorVotes,
}

impl AllowedComponents {
    pub fn from_str(s: &str) -> Result<Self, Diagnostic> {
        match s {
            "Account" => Ok(AllowedComponents::Account),
            "EthAccount" => Ok(AllowedComponents::EthAccount),
            "SRC9" => Ok(AllowedComponents::SRC9),
            "AccessControl" => Ok(AllowedComponents::AccessControl),
            "Ownable" => Ok(AllowedComponents::Ownable),
            "Vesting" => Ok(AllowedComponents::Vesting),
            "SRC5" => Ok(AllowedComponents::SRC5),
            "Initializable" => Ok(AllowedComponents::Initializable),
            "Pausable" => Ok(AllowedComponents::Pausable),
            "ReentrancyGuard" => Ok(AllowedComponents::ReentrancyGuard),
            "ERC20" => Ok(AllowedComponents::ERC20),
            "ERC721" => Ok(AllowedComponents::ERC721),
            "ERC721Enumerable" => Ok(AllowedComponents::ERC721Enumerable),
            "ERC721Receiver" => Ok(AllowedComponents::ERC721Receiver),
            "ERC1155" => Ok(AllowedComponents::ERC1155),
            "ERC1155Receiver" => Ok(AllowedComponents::ERC1155Receiver),
            "ERC2981" => Ok(AllowedComponents::ERC2981),
            "ERC4626" => Ok(AllowedComponents::ERC4626),
            "Upgradeable" => Ok(AllowedComponents::Upgradeable),
            "Nonces" => Ok(AllowedComponents::Nonces),
            "Multisig" => Ok(AllowedComponents::Multisig),
            "TimelockController" => Ok(AllowedComponents::TimelockController),
            "Votes" => Ok(AllowedComponents::Votes),
            "Governor" => Ok(AllowedComponents::Governor),
            "GovernorCoreExecution" => Ok(AllowedComponents::GovernorCoreExecution),
            "GovernorCountingSimple" => Ok(AllowedComponents::GovernorCountingSimple),
            "GovernorSettings" => Ok(AllowedComponents::GovernorSettings),
            "GovernorTimelockExecution" => Ok(AllowedComponents::GovernorTimelockExecution),
            "GovernorVotesQuorumFraction" => Ok(AllowedComponents::GovernorVotesQuorumFraction),
            "GovernorVotes" => Ok(AllowedComponents::GovernorVotes),
            _ => Err(Diagnostic::error(errors::INVALID_COMPONENT(s))),
        }
    }

    pub fn get_info(&self) -> ComponentInfo<'static> {
        match self {
            AllowedComponents::Account => ComponentInfo {
                name: "AccountComponent",
                path: "openzeppelin_account::AccountComponent",
                storage: "account",
                event: "AccountEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::EthAccount => ComponentInfo {
                name: "EthAccountComponent",
                path: "openzeppelin_account::EthAccountComponent",
                storage: "eth_account",
                event: "EthAccountEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::SRC9 => ComponentInfo {
                name: "SRC9Component",
                path: "openzeppelin_account::extensions::SRC9Component",
                storage: "src9",
                event: "SRC9Event",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::Ownable => ComponentInfo {
                name: "OwnableComponent",
                path: "openzeppelin_access::ownable::OwnableComponent",
                storage: "ownable",
                event: "OwnableEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::AccessControl => ComponentInfo {
                name: "AccessControlComponent",
                path: "openzeppelin_access::accesscontrol::AccessControlComponent",
                storage: "access_control",
                event: "AccessControlEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::Vesting => ComponentInfo {
                name: "VestingComponent",
                path: "openzeppelin_finance::vesting::VestingComponent",
                storage: "vesting",
                event: "VestingEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::SRC5 => ComponentInfo {
                name: "SRC5Component",
                path: "openzeppelin_introspection::src5::SRC5Component",
                storage: "src5",
                event: "SRC5Event",
                has_initializer: false,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::Initializable => ComponentInfo {
                name: "InitializableComponent",
                path: "openzeppelin_security::InitializableComponent",
                storage: "initializable",
                event: "InitializableEvent",
                has_initializer: false,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::Pausable => ComponentInfo {
                name: "PausableComponent",
                path: "openzeppelin_security::PausableComponent",
                storage: "pausable",
                event: "PausableEvent",
                has_initializer: false,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::ReentrancyGuard => ComponentInfo {
                name: "ReentrancyGuardComponent",
                path: "openzeppelin_security::ReentrancyGuardComponent",
                storage: "reentrancy_guard",
                event: "ReentrancyGuardEvent",
                has_initializer: false,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::ERC20 => ComponentInfo {
                name: "ERC20Component",
                path: "openzeppelin_token::erc20::ERC20Component",
                storage: "erc20",
                event: "ERC20Event",
                has_initializer: true,
                has_immutable_config: true,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::ERC721 => ComponentInfo {
                name: "ERC721Component",
                path: "openzeppelin_token::erc721::ERC721Component",
                storage: "erc721",
                event: "ERC721Event",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::ERC721Enumerable => ComponentInfo {
                name: "ERC721EnumerableComponent",
                path: "openzeppelin_token::erc721::extensions::ERC721EnumerableComponent",
                storage: "erc721_enumerable",
                event: "ERC721EnumerableEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::ERC721Receiver => ComponentInfo {
                name: "ERC721ReceiverComponent",
                path: "openzeppelin_token::erc721::ERC721ReceiverComponent",
                storage: "erc721_receiver",
                event: "ERC721ReceiverEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::ERC1155 => ComponentInfo {
                name: "ERC1155Component",
                path: "openzeppelin_token::erc1155::ERC1155Component",
                storage: "erc1155",
                event: "ERC1155Event",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::ERC1155Receiver => ComponentInfo {
                name: "ERC1155ReceiverComponent",
                path: "openzeppelin_token::erc1155::ERC1155ReceiverComponent",
                storage: "erc1155_receiver",
                event: "ERC1155ReceiverEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::ERC2981 => ComponentInfo {
                name: "ERC2981Component",
                path: "openzeppelin_token::common::erc2981::ERC2981Component",
                storage: "erc2981",
                event: "ERC2981Event",
                has_initializer: true,
                has_immutable_config: true,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::ERC4626 => ComponentInfo {
                name: "ERC4626Component",
                path: "openzeppelin_token::erc20::extensions::erc4626::ERC4626Component",
                storage: "erc4626",
                event: "ERC4626Event",
                has_initializer: true,
                has_immutable_config: true,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::Upgradeable => ComponentInfo {
                name: "UpgradeableComponent",
                path: "openzeppelin_upgrades::UpgradeableComponent",
                storage: "upgradeable",
                event: "UpgradeableEvent",
                has_initializer: false,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::Nonces => ComponentInfo {
                name: "NoncesComponent",
                path: "openzeppelin_utils::nonces::NoncesComponent",
                storage: "nonces",
                event: "NoncesEvent",
                has_initializer: false,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::Multisig => ComponentInfo {
                name: "MultisigComponent",
                path: "openzeppelin_governance::multisig::MultisigComponent",
                storage: "multisig",
                event: "MultisigEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::TimelockController => ComponentInfo {
                name: "TimelockControllerComponent",
                path: "openzeppelin_governance::timelock::TimelockControllerComponent",
                storage: "timelock_controller",
                event: "TimelockControllerEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::Votes => ComponentInfo {
                name: "VotesComponent",
                path: "openzeppelin_governance::votes::VotesComponent",
                storage: "votes",
                event: "VotesEvent",
                has_initializer: false,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::Governor => ComponentInfo {
                name: "GovernorComponent",
                path: "openzeppelin_governance::governor::GovernorComponent",
                storage: "governor",
                event: "GovernorEvent",
                has_initializer: true,
                has_immutable_config: true,
                internal_impls: vec!["InternalImpl"],
            },
            AllowedComponents::GovernorCoreExecution => ComponentInfo {
                name: "GovernorCoreExecutionComponent",
                path: "openzeppelin_governance::governor::extensions::GovernorCoreExecutionComponent",
                storage: "governor_core_execution",
                event: "GovernorCoreExecutionEvent",
                has_initializer: false,
                has_immutable_config: false,
                internal_impls: vec!["GovernorExecution"],
            },
            AllowedComponents::GovernorCountingSimple => ComponentInfo {
                name: "GovernorCountingSimpleComponent",
                path: "openzeppelin_governance::governor::extensions::GovernorCountingSimpleComponent",
                storage: "governor_counting_simple",
                event: "GovernorCountingSimpleEvent",
                has_initializer: false,
                has_immutable_config: false,
                internal_impls: vec!["GovernorCounting"],
            },
            AllowedComponents::GovernorSettings => ComponentInfo {
                name: "GovernorSettingsComponent",
                path: "openzeppelin_governance::governor::extensions::GovernorSettingsComponent",
                storage: "governor_settings",
                event: "GovernorSettingsEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl", "GovernorSettings"],
            },
            AllowedComponents::GovernorTimelockExecution => ComponentInfo {
                name: "GovernorTimelockExecutionComponent",
                path: "openzeppelin_governance::governor::extensions::GovernorTimelockExecutionComponent",
                storage: "governor_timelock_execution",
                event: "GovernorTimelockExecutionEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl", "GovernorExecution"],
            },
            AllowedComponents::GovernorVotesQuorumFraction => ComponentInfo {
                name: "GovernorVotesQuorumFractionComponent",
                path: "openzeppelin_governance::governor::extensions::GovernorVotesQuorumFractionComponent",
                storage: "governor_votes_quorum_fraction",
                event: "GovernorVotesQuorumFractionEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl", "GovernorQuorum", "GovernorVotes"],
            },
            AllowedComponents::GovernorVotes => ComponentInfo {
                name: "GovernorVotesComponent",
                path: "openzeppelin_governance::governor::extensions::GovernorVotesComponent",
                storage: "governor_votes",
                event: "GovernorVotesEvent",
                has_initializer: true,
                has_immutable_config: false,
                internal_impls: vec!["InternalImpl", "GovernorVotes"],
            },
        }
    }
}

/// Information about a component.
///
/// # Members
///
/// * `name` - The name of the component (e.g. `ERC20Component`)
/// * `path` - The path from where the component is imported (e.g. `openzeppelin_token::erc20::ERC20Component`)
/// * `storage` - The path to reference the component in storage (e.g. `erc20`)
/// * `event` - The path to reference the component events (e.g. `ERC20Event`)
/// * `has_initializer` - Whether the component requires an initializer (e.g. `true`)
/// * `internal_impls` - The internal implementations of the component to be added to
///   the module by default (e.g. `["InternalImpl1", "InternalImpl2"]`)
#[derive(Debug, Clone)]
pub struct ComponentInfo<'a> {
    pub name: &'a str,
    pub path: &'a str,
    pub storage: &'a str,
    pub event: &'a str,
    pub has_initializer: bool,
    pub has_immutable_config: bool,
    pub internal_impls: Vec<&'a str>,
}

impl<'a> ComponentInfo<'a> {
    pub fn short_name(&self) -> &'a str {
        self.name
            .split("Component")
            .next()
            .expect("Component name must end with 'Component'")
    }

    pub fn kind(&self) -> AllowedComponents {
        // It should be safe to unwrap here since the ComponentInfo must have
        // been created from a valid AllowedComponent
        AllowedComponents::from_str(self.short_name()).unwrap()
    }
}
