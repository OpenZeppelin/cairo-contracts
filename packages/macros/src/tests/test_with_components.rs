use crate::with_components::with_components_avevetedp5blk as with_components;
use cairo_lang_macro::TokenStream;
use indoc::{formatdoc, indoc};
use insta::assert_snapshot;

#[test]
fn test_with_erc20() {
    let attribute = "(ERC20)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyToken {
            use openzeppelin_token::erc20::ERC20HooksEmptyImpl;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
                self.erc20.initializer(\"MyToken\", \"MTK\");
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_erc20_no_initializer() {
    let attribute = "(ERC20)";
    let item = indoc!(
        "
      #[starknet::contract]
      pub mod MyToken {
          use openzeppelin_token::erc20::ERC20HooksEmptyImpl;
          use starknet::ContractAddress;

          #[storage]
          pub struct Storage {}

          #[constructor]
          fn constructor(ref self: ContractState) {
          }
      }
      "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_ownable() {
    let attribute = "(Ownable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod Owned {
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState, owner: ContractAddress) {
                self.ownable.initializer(owner);
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_ownable_no_initializer() {
    let attribute = "(Ownable)";
    let item = indoc!(
        "
      #[starknet::contract]
      pub mod Owned {
          use starknet::ContractAddress;

          #[storage]
          pub struct Storage {}

          #[constructor]
          fn constructor(ref self: ContractState) {
          }
      }
      "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_two_components() {
    let attribute = "(ERC20, Ownable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyToken {
            use openzeppelin_token::erc20::ERC20HooksEmptyImpl;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState, owner: ContractAddress) {
                self.ownable.initializer(owner);
                self.erc20.initializer(\"MyToken\", \"MTK\");
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_two_components_no_initializer() {
    let attribute = "(ERC20, Ownable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyToken {
            use openzeppelin_token::erc20::ERC20HooksEmptyImpl;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState) {
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_two_components_no_constructor() {
    let attribute = "(ERC20, Ownable)";
    let item = indoc!(
        "
        #[starknet::contract]
        pub mod MyToken {
            use openzeppelin_token::erc20::ERC20HooksEmptyImpl;
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_no_contract_attribute() {
    let attribute = "(Ownable)";
    let item = indoc!(
        "
        pub mod Owned {
            use starknet::ContractAddress;

            #[storage]
            pub struct Storage {}

            #[constructor]
            fn constructor(ref self: ContractState, owner: ContractAddress) {
                self.ownable.initializer(owner);
            }
        }
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

#[test]
fn test_with_no_body() {
    let attribute = "(ERC20, Ownable)";
    let item = indoc!(
        "
        pub mod MyToken;
        "
    );
    let result = get_string_result(attribute, item);
    assert_snapshot!(result);
}

//
// Helpers
//

/// Returns a string representation of the result of the macro expansion,
/// including the token stream, diagnostics and aux data.
fn get_string_result(attribute: &str, item: &str) -> String {
    let attribute_stream = TokenStream::new(attribute.to_string());
    let item_stream = TokenStream::new(item.to_string());
    let raw_result = with_components(attribute_stream, item_stream);
    let none = "None";

    let mut token_stream = raw_result.token_stream.to_string();
    let mut diagnostics = String::new();
    for d in raw_result.diagnostics {
        diagnostics += format!("====\n{:?}: {}====", d.severity, d.message).as_str();
    }

    if token_stream.is_empty() {
        token_stream = none.to_string();
    }
    if diagnostics.is_empty() {
        diagnostics = none.to_string();
    }

    formatdoc! {
        "
        TokenStream:

        {}

        Diagnostics:

        {}

        AuxData:

        {:?}
        ",
        token_stream, diagnostics, raw_result.aux_data
    }
}
