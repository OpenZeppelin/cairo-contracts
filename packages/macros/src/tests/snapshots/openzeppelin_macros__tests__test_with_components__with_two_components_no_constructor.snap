---
source: src/tests/test_with_components.rs
expression: result
snapshot_kind: text
---
TokenStream:

#[starknet::contract]
pub mod MyToken {
    use openzeppelin_token::erc20::ERC20HooksEmptyImpl;
    use starknet::ContractAddress;

    #[storage]
    pub struct Storage {        #[substorage(v0)]
        erc20: ERC20Component::Storage,
        #[substorage(v0)]
        ownable: OwnableComponent::Storage,
}

    use openzeppelin_token::erc20::ERC20Component;
    use openzeppelin_access::ownable::OwnableComponent;

    component!(path: ERC20Component, storage: erc20, event: ERC20Event);
    component!(path: OwnableComponent, storage: ownable, event: OwnableEvent);

    impl ERC20InternalImpl = ERC20Component::InternalImpl<ContractState>;
    impl OwnableInternalImpl = OwnableComponent::InternalImpl<ContractState>;

    #[event]
    #[derive(Drop, starknet::Event)]
    enum Event {
        #[flat]
        ERC20Event: ERC20Component::Event,
        #[flat]
        OwnableEvent: OwnableComponent::Event,
    }}


Diagnostics:

====
Warning: It looks like the initilizers for the following components are missing:

ERC20, Ownable

This may lead to unexpected behavior. We recommend adding a constructor with the corresponding initializer calls.
====

AuxData:

None
